types/retry-handler.d.ts

```diff
@@ -0,0 +1,109 @@
+import Dispatcher from "./dispatcher";
+
+export default RetryHandler;
+
+declare class RetryHandler implements Dispatcher.DispatchHandlers {
+  constructor(
+    options: Dispatcher.DispatchOptions,
+    retryHandlers: RetryHandler.RetryHandlers,
+    retryOptions?: RetryHandler.RetryOptions
+  );
+}
+
+declare namespace RetryHandler {
+  export type RetryState = { counter: number; currentTimeout: number };
+
+  export type RetryCallback = (
+    err: Error,
+    state: RetryState,
+    opts: RetryOptions
+  ) => number | null;
+
+  export interface RetryOptions {
+    /**
+     * Callback to be invoked on every retry iteration.
+     * It receives the error, current state of the retry object and the options object
+     * passed when instantiating the retry handler.
+     *
+     * @type {RetryCallback}
+     * @memberof RetryOptions
+     */
+    retry?: RetryCallback;
+    /**
+     * Maximum number of retries to allow.
+     *
+     * @type {number}
+     * @memberof RetryOptions
+     * @default 5
+     */
+    max?: number;
+    /**
+     * Max number of milliseconds allow between retries
+     *
+     * @type {number}
+     * @memberof RetryOptions
+     * @default 30000
+     */
+    maxTimeout?: number;
+    /**
+     * Initial number of milliseconds to wait before retrying for the first time.
+     *
+     * @type {number}
+     * @memberof RetryOptions
+     * @default 500
+     */
+    minTimeout?: number;
+    /**
+     * Factior to multiply the timeout factor between retries.
+     *
+     * @type {number}
+     * @memberof RetryOptions
+     * @default 2
+     */
+    timeoutFactor?: number;
+    /**
+     * It enables to automatically infer timeout between retries based on the `Retry-After` header.
+     *
+     * @type {boolean}
+     * @memberof RetryOptions
+     * @default true
+     */
+    retryAfter?: boolean;
+    /**
+     * HTTP methods to retry.
+     *
+     * @type {Dispatcher.HttpMethod[]}
+     * @memberof RetryOptions
+     */
+    methods?: Dispatcher.HttpMethod[];
+    /**
+     * It enables automatic retry between requets on idempotent methods.
+     * The method should be defined as safe accordingly to https://developer.mozilla.org/en-US/docs/Glossary/Safe/HTTP
+     *
+     * Note: It takes presedence over the `methods` option.
+     * 
+     * @type {boolean}
+     * @memberof RetryOptions
+     * @default false
+     */
+    idempotent?: boolean;
```

Comment:
Shouldn't this be true by default?

Reply:
Have some conflicts deciding that; especially as the `methods` is also enabled by default, so it might make sense have one of them enabled. Maybe push for `idempotent` and disabled `methods`?

Reply:
Hm, since there is overlap, maybe we can make it so methods is either an array of methods or the string `'idempotent'`?

Reply:
Not sure, maybe easiest to just remove this option?

Reply:
I'd say we better remove it, it will make our lifes and implementers easier. Anyways, if they want to automatically retry on certain method, they can set it in the `method` array